src/components/header/EmptyThreadIndicator.js

```diff
@@ -126,4 +75,72 @@ class EmptyThreadIndicator extends PureComponent<Props> {
   }
 }
 
+/**
+ * Define this outside of the class so that it's easily testable. The internals
+ * are a little complicated with the math, but the test file should have some
+ * pretty clear explanations of the requirements:
+ * src/test/components/EmptyThreadIndicator.test.js
+ */
+export function getIndicatorPositions(
+  props: Props
+): {|
+  startup: SyntheticCssDeclarations | null,
+  shutdown: SyntheticCssDeclarations | null,
+  emptyBufferStart: SyntheticCssDeclarations | null,
+|} {
+  const {
+    rangeStart,
+    rangeEnd,
+    width,
+    unfilteredSamplesRange,
+    interval,
+    thread: { processShutdownTime, registerTime, unregisterTime },
+  } = props;
+  const rangeLength = rangeEnd - rangeStart;
+  const xPixelsPerMs = width / rangeLength;
+  const threadEndTime =
+    unregisterTime === null ? processShutdownTime : unregisterTime;
+
+  // Did this thread startup in this time range?
+  let startup = null;
+  if (registerTime > rangeStart) {
+    startup = {
+      left: 0,
+      width: Math.min(width, (registerTime - rangeStart) * xPixelsPerMs),
+    };
+  }
+
+  // Did this thread shut down in this time range?
+  let shutdown = null;
+  if (threadEndTime !== null && threadEndTime < rangeEnd) {
+    shutdown = {
+      right: 0,
+      width: Math.min(width, (rangeEnd - threadEndTime) * xPixelsPerMs),
+    };
+  }
+
+  let emptyBufferStart = null;
+  if (
+    // Threads could have no samples, and therefore no range.
+    unfilteredSamplesRange !== null &&
+    // Was the buffer empty at the beginning of the range, at least one interval length
+    // into the profile? This interval length ensures no awkward cut off where it's not
+    // really needed.
+    unfilteredSamplesRange.start > rangeStart + interval &&
+    // Only show this if it's actually in the current range.
+    registerTime < rangeEnd
+  ) {
+    const startMilliseconds = Math.max(0, registerTime - rangeStart);
+    emptyBufferStart = {
+      left: startMilliseconds * xPixelsPerMs,
+      width:
+        (Math.min(unfilteredSamplesRange.start, rangeEnd) -
+          Math.max(registerTime, rangeStart)) *
+        xPixelsPerMs,
```

Comment:
So I tried various variations for this calculation, especially because I wanted the same "template" with `Math.min(width, ...)` than in the other calculations above.

Variation 1:
```javascript
Math.min(width, (unfilteredSamplesRange.start - registerTime) * xPixelsPerMs)
```

From my tests, this doesn't work when a buffer is rewritten (example profile: http://localhost:4242/public/116d028e3e2e1e3c1e4b5ade04ee55ff1568f670/calltree/?hiddenThreads=&react_perf&thread=2&threadOrder=0-2-3-4-5-1&v=2).
_But_ the tests still pass with the change so I think you miss a case in the tests.

But my variation 2 works:
```javascript
      width: Math.min(
        width,
        (unfilteredSamplesRange.start - Math.max(registerTime, rangeStart)) *
          xPixelsPerMs
      ),
```
This isn't very different that what you wrote though, although it's more consistent with the other calculations. So I'll leave it to yout judgment.

Reply:
I'll try to elaborate a bit more.

I tried to change the calculation you have in the code with the one in Variation 1 above, that I thought was a simpler solution. The tests were green, so I thought it worked. But visually I see it doesnt work on that profile. That's why I say a test is missing, to fail in a case similar to that profile.

I hope this is clearer ?

Reply:
Got this updated and added 2 new test cases that I think only apply to the `emptyBufferStart` case.